### PR TITLE
feat: Single-submission per user (optional gate for private forms)

### LINKED ISSUE #580: Single Submission Enforcement + Test Submission Flag Across API and Hub
### Summary
Implement single-submission enforcement for private forms with a controlled tester bypass flow, and expose duplicate-submission state to public form UIs so respondents see an `Already Responded` experience instead of the survey when appropriate.

### Problem
Some use cases require protection of survey data quality by preventing duplicate responses from the same authenticated user on private forms. At the same time, QA/demo users need a safe way to submit multiple times without polluting production reporting.

### Proposed Solution
- Add a form rule to limit submissions to one per user on private forms.
- Enforce duplicate blocking in API create-submission flow.
- Introduce `forms.test` permission to bypass duplicate blocking for authorized testers.
- Persist `IsTestSubmission` on bypassed submissions.
- Expose `HasUserSubmitted` via active-definition APIs.
- Gate public `share` and `embed` Hub flows to render `AlreadyResponded` for non-token duplicate responders.
- Keep token edit/resume flows working (do not gate with `AlreadyResponded` when valid token is present).

### Scope
- **Backend (API)**
  - Form/submission contract updates (`LimitOnePerUser`, metadata, `IsTestSubmission`)
  - Duplicate check and `409 Conflict` behavior
  - `Actions.Forms.Test` permission and bypass handling
  - Active definition response includes `HasUserSubmitted`
- **Frontend (Hub)**
  - Public form gating (`share` + `embed`) based on `HasUserSubmitted`
  - Reusable `AlreadyResponded` component with metadata-driven message
  - Preserve embed iframe resizing behavior for blocked state

### Acceptance Criteria
1. Private form + `LimitOnePerUser=true` blocks non-test duplicate submissions (`409`).
2. Users with `forms.test` can submit repeatedly and submissions are flagged as test.
3. Active definition includes `HasUserSubmitted` for relevant flows.
4. Public `share`/`embed` routes show `AlreadyResponded` instead of survey when user already submitted and no edit token is present.
5. Token-based edit/resume flows are not blocked by duplicate-gating UI.
6. `AlreadyResponded` message supports metadata config with safe fallback defaults.

### Notes
- Keep API changes additive and non-breaking where possible.
- Ensure parity between create-submission enforcement and active-definition gating logic.

### PR DESCRIPTION
**Pull request by orilux**
## Cumulative PR

- https://github.com/endatix/endatix/pull/710
- https://github.com/endatix/endatix/pull/713
- https://github.com/endatix/endatix/pull/716

## Related Issues
https://github.com/endatix/endatix-hub/issues/580

## Type of Change
Check all that apply.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Dependency update
- [ ] Refactoring (non-breaking change which improves code quality)
- [ ] Other (please describe)

## Checklist
- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- New and existing tests pass locally with my changes
- Any dependent changes have been merged and published in downstream modules

> [!NOTE]
> - [x] I confirm that my Pull Request follows all the checklist requirements above

### CODE REVIEW

File: src/Endatix.Api/Endpoints/Forms/JsonStringValidationExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+internal static class JsonStringValidationExtensions
+{
+    internal static IRuleBuilderOptions<T, string?> MustBeValidJsonString<T>(this IRuleBuilder<T, string?> ruleBuilder)
+    {
+        return ruleBuilder
+            .Must(JsonStringValidation.IsValid)
+            .WithMessage("Metadata must be a valid JSON string.");
+    }
+}
```
Comment (PeterMarinov) [inline src/Endatix.Api/Endpoints/Forms/JsonStringValidationExtensions.cs:13]: 2 opportunities with this method:
1. We must move out of Endatix.Api.Endpoints.Forms => Endatix.Api.Endpoints.Common so it's for all Api endpionts and make it public so other Devs can use it
2. Make the Message more contexual so instead of "Metadata" it shows "{ParamaterName}"

Added a task to address this so we don't slow the PR - https://github.com/endatix/endatix/issues/718

File: src/Endatix.Api/Endpoints/Forms/PartialUpdate.cs [hunk 2]
```diff
@@ -30,14 +32,28 @@ public override void Configure()
...
+                Metadata = request.Metadata
+            },
             cancellationToken);
 
-        return TypedResultsBuilder
-            .MapResult(result, FormMapper.Map<PartialUpdateFormResponse>)
-            .SetTypedResults<Ok<PartialUpdateFormResponse>, BadRequest, NotFound>();
+        var mappedResult = result.Map(FormMapper.Map<PartialUpdateFormResponse>);
+
+        return mappedResult.Status switch
```
Comment (PeterMarinov) [inline src/Endatix.Api/Endpoints/Forms/PartialUpdate.cs:53]: Just FYI:

We would use  TypedResultsMapper, which does all the heavy lifting and the messages propagation
```cs
return TypedResultsBuilder
            .MapResult(result, FormMapper.Map<PartialUpdateFormResponse>)
            .SetTypedResults<Ok<PartialUpdateFormResponse>, ProblemHttpResult>();
```

File: src/Endatix.Api/Endpoints/Forms/Update.cs [hunk 3]
```diff
@@ -30,14 +32,25 @@ public override void Configure()
...
+                request.LimitOnePerUser,
+                request.Metadata),
             cancellationToken);
 
-        return TypedResultsBuilder
-            .MapResult(result, FormMapper.Map<UpdateFormResponse>)
-            .SetTypedResults<Ok<UpdateFormResponse>, BadRequest, NotFound>();
+        var mappedResult = result.Map(FormMapper.Map<UpdateFormResponse>);
+
+        return mappedResult.Status switch
```
Comment (PeterMarinov) [inline src/Endatix.Api/Endpoints/Forms/Update.cs:50]: Same here
```cs
return TypedResultsBuilder
            .MapResult(result, FormMapper.Map<UpdateFormResponse>)
            .SetTypedResults<Ok<UpdateFormResponse>, ProblemHttpResult>();
``

File: src/Endatix.Core/UseCases/Forms/Update/UpdateFormHandler.cs [hunk 4]
```diff
@@ -3,14 +3,20 @@
...
 namespace Endatix.Core.UseCases.Forms.Update;
 
 public class UpdateFormHandler(
     IRepository<Form> repository,
+    IRepository<Submission> submissionRepository,
     IMediator mediator) : ICommandHandler<UpdateFormCommand, Result<Form>>
 {
+    private const string ENABLE_CONFLICT_MESSAGE = "Cannot enable single submission gate because this form already has duplicate submissions.";
+    private const string DISABLE_CONFLICT_MESSAGE = "Single submission gate cannot be disabled after it has been enabled.";
+    private const string PUBLIC_CONFLICT_MESSAGE = "A single-submission form cannot be made public.";
```
Comment (PeterMarinov) [inline src/Endatix.Core/UseCases/Forms/Update/UpdateFormHandler.cs:18]: Maybe message should be extended to also say. "First make private and try again"?

File: src/Endatix.Core/UseCases/Forms/Update/UpdateFormHandler.cs [hunk 5]
```diff
@@ -20,9 +26,35 @@ public async Task<Result<Form>> Handle(UpdateFormCommand request, CancellationTo
...
+        }
+
+        if (form.IsPublic && requestedLimitOnePerUser)
+        {
+            return Result<Form>.Conflict(PUBLIC_CONFLICT_MESSAGE);
+        }
+
+        if (!form.LimitOnePerUser && requestedLimitOnePerUser)
+        {
+            var hasDuplicateEligibleSubmissions = await HasDuplicateEligibleSubmissionsAsync(
```
Comment (PeterMarinov) [inline src/Endatix.Core/UseCases/Forms/Update/UpdateFormHandler.cs:43]: This is rare case, but we might have an issue that the soft check says "No existing duplicated submissions and after the read, the duplication occurs". One solution might be to update submissions with the RestrictedIndex value and ensure the duplicaiton is not happening on DB level. Another solution might be to prevent changing the type of submission at all - e.g. enforce it on form creation and then lock the value to read-only. I think, this will simplify data integrity and our logic to ensure it or we will have more and more edge cases like this to handle, test and we will be bound to error prone use cases

File: src/Endatix.Core/Specifications/EligibleSingleSubmissionGateSubmitterIdsByFormIdSpec.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
+using Ardalis.Specification;
+using Endatix.Core.Entities;
+
+namespace Endatix.Core.Specifications;
+
+public sealed class EligibleSingleSubmissionGateSubmitterIdsByFormIdSpec : Specification<Submission, string>
+{
+    public EligibleSingleSubmissionGateSubmitterIdsByFormIdSpec(long formId)
```
Comment (PeterMarinov) [inline src/Endatix.Core/Specifications/EligibleSingleSubmissionGateSubmitterIdsByFormIdSpec.cs:8]: Just sharing thoughts. We should focus the name on the data access flow, not the business context as this is pure data access concern.

Also, I would shorten the name as this is too long for reading, writing and understanding => something like "UserSignedSubmissionsByFormIdSpec"

File: src/Endatix.Core/UseCases/Forms/SingleSubmissionGateDuplicateChecker.cs [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+namespace Endatix.Core.UseCases.Forms;
+
+internal static class SingleSubmissionGateDuplicateChecker
+{
+    public static async Task<bool> HasDuplicateEligibleSubmissionsAsync(
+        long formId,
+        IRepository<Submission> submissionRepository,
+        CancellationToken cancellationToken)
+    {
+        var eligibleSubmitterIds = await submissionRepository.ListAsync(
```
Comment (PeterMarinov) [inline src/Endatix.Core/UseCases/Forms/SingleSubmissionGateDuplicateChecker.cs:14]: imo _existing_ seems more suitable than _eligible_ in this context

File: src/Endatix.Core/UseCases/Forms/SingleSubmissionGateDuplicateChecker.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+            new EligibleSingleSubmissionGateSubmitterIdsByFormIdSpec(formId),
+            cancellationToken);
+
+        return eligibleSubmitterIds
```
Comment (PeterMarinov) [inline src/Endatix.Core/UseCases/Forms/SingleSubmissionGateDuplicateChecker.cs:18]: This is potential issue. We load all submittedByIds in memory and then do the grouping to identify submittedBy with more than one submissions.

We should do all of that in the DB e.g. 

```cs
public HasMoreThanOneUserSignedSubmissionByFormIdSpec(long formId)
    {
        Query.Where(s =>
               s.FormId == formId &&
               s.SubmittedBy != null &&
               s.SubmittedBy != string.Empty &&
               !s.IsTestSubmission)
         .GroupBy(s => s.SubmittedBy)
         .AnyAsync(g => g.Count() > 1);
```

File: src/Endatix.Core/UseCases/Submissions/Create/CreateSubmissionHandler.cs [hunk 9]
```diff
@@ -55,6 +57,23 @@ public async Task<Result<Submission>> Handle(CreateSubmissionCommand request, Ca
             }
         }
 
+        var canBypassSingleSubmissionLimit = false;
```
Comment (PeterMarinov) [inline src/Endatix.Core/UseCases/Submissions/Create/CreateSubmissionHandler.cs:60]: We do check for "test" permission only based of the form settings e.g. the form decides, not the user

Won't be correct if instead we check for the permission only if the user request to create a test permission. Imo will be cleaner in terms of code structure (less defensive code, more precise and controlled flow) and more performant Create - e.g.  auth check is done only if Test permission is requested.

I haven't check the error propagation but now it it return Forbidden that "Actions.Forms.Test" is required even if the users are just regular users trying to submit more than one response
1. Regular users -> see conflcit e.g. you cannot have more than one submission
2. Users requesting test submission -> "You cannot create test submission. `Actions.Forms.Test` is required "

File: src/Endatix.Infrastructure/Data/AppDbContext.cs [hunk 10]
```diff
@@ -83,15 +86,83 @@ protected override void OnModelCreating(ModelBuilder builder)
...
-        ProcessEntities();
-        return base.SaveChanges();
+        try
+        {
+            ProcessEntities();
+            return base.SaveChanges();
+        }
+        catch (DbUpdateException dbUpdateException) when (IsDuplicateSubmissionConstraintViolation(dbUpdateException))
+        {
+            throw new DuplicateSubmissionException("A submission already exists for this user and form.", dbUpdateException);
```
Comment (PeterMarinov) [inline src/Endatix.Infrastructure/Data/AppDbContext.cs:96]: Just a note that this looks like something we must offload to Submission handlers and repositories as it's Submission specific and keep DbContext clean - will try to fix later

**Review by PeterMarinov — DISMISSED** [review]
Approved. Great addition of the Options pattern for the juicy Entity constructors to separate required from optional params. 

Please read the comments before merging and let's discuss so we can capture any future work

File: src/Endatix.Api/Endpoints/Forms/JsonStringValidationExtensions.cs (hunk 1 above)
Comment (orilux) [reply src/Endatix.Api/Endpoints/Forms/JsonStringValidationExtensions.cs:13]: noted

File: src/Endatix.Api/Endpoints/Forms/PartialUpdate.cs (hunk 2 above)
Comment (orilux) [reply src/Endatix.Api/Endpoints/Forms/PartialUpdate.cs:53]: noted

File: src/Endatix.Api/Endpoints/Forms/Update.cs (hunk 3 above)
Comment (orilux) [reply src/Endatix.Api/Endpoints/Forms/Update.cs:50]: ditto

File: src/Endatix.Core/UseCases/Forms/Update/UpdateFormHandler.cs (hunk 4 above)
Comment (orilux) [reply src/Endatix.Core/UseCases/Forms/Update/UpdateFormHandler.cs:18]: This message explains why the "Public" button is disabled. Either way, this will go away once we commit to form types.

File: src/Endatix.Core/UseCases/Forms/Update/UpdateFormHandler.cs (hunk 5 above)
Comment (orilux) [reply src/Endatix.Core/UseCases/Forms/Update/UpdateFormHandler.cs:43]: This was also flagged by AI tools, but I decided against premature optimization on it because:
- It is extremely unlikely for a real world form that has collected hundreds of submission in multi-submitter mode to get switched to "single"
- We will probably remove that altogether when we commit to form types.

File: src/Endatix.Core/Specifications/EligibleSingleSubmissionGateSubmitterIdsByFormIdSpec.cs (hunk 6 above)
Comment (orilux) [reply src/Endatix.Core/Specifications/EligibleSingleSubmissionGateSubmitterIdsByFormIdSpec.cs:8]: noted

File: src/Endatix.Core/UseCases/Forms/SingleSubmissionGateDuplicateChecker.cs (hunk 7 above)
Comment (orilux) [reply src/Endatix.Core/UseCases/Forms/SingleSubmissionGateDuplicateChecker.cs:14]: "Eligible" refers to them not being test submissions.

File: src/Endatix.Core/UseCases/Forms/SingleSubmissionGateDuplicateChecker.cs (hunk 8 above)
Comment (orilux) [reply src/Endatix.Core/UseCases/Forms/SingleSubmissionGateDuplicateChecker.cs:18]: I will look into this.

File: src/Endatix.Core/UseCases/Submissions/Create/CreateSubmissionHandler.cs (hunk 9 above)
Comment (orilux) [reply src/Endatix.Core/UseCases/Submissions/Create/CreateSubmissionHandler.cs:60]: We have a very basic test-submission capability at this time. We should consider expanding this into allowing users to enable "test mode" on demand.

File: src/Endatix.Infrastructure/Data/AppDbContext.cs (hunk 10 above)
Comment (orilux) [reply src/Endatix.Infrastructure/Data/AppDbContext.cs:96]: Ok.

**Review by PeterMarinov — APPROVED** [review]
Approved
